Add EVP_PKEY_get_group_name and use that in libssl instead of evp_pkey_get_EC_KEY_curve_nid()

levitte opened the pull request:
evp_pkey_get_EC_KEY_curve_nid() is an internal function solely used by libssl, and needs refactoring while not transporting legacy key handling to libssl.

The result is the addition of `EVP_PKEY_get_group_name()`, which is inspired by the operation function `EVP_PKEY_CTX_get_group_name()`, and which can be used in libssl.  This is currently done with an internal libssl function that converts the group name to a NID, since that's still used in numerous libssl locations.

Future libssl development includes to change those locations to use group names directly instead, and possibly provider capabilities.  That is, however, for another PR.

Alternative to #13371

paulidale reviewed (commented):
I expected documentation and test changes/additions.

levitte commented:
> I expected documentation

Sure.  If someone documents `EC_curve_nist2nid()`, I'll tag along with an added `EC_curve_name2nid()` description.
I am *not* knowledgable enough re EC to write a good description from scratch.

> and test changes/additions.

Why?  This is not *new* functionality, it's only moved.  In other words, the tests that include a call to `evp_pkey_get_EC_KEY_curve_nid()` now include a call to `ssl_get_EC_curve_nid()`.

paulidale commented:
I'd argue that this isn't just moved functionality.  There is a new _public_ function, this needs to be documented for 3.0 beta 1.

levitte commented:
> I'd argue that this isn't just moved functionality.  There is a new _public_ function, this needs to be documented for 3.0 beta 1.

Yes?  Please read again what I wrote about documentation

levitte commented:
Ping?

levitte commented:
Ping!

levitte commented:
A concern has been made re documentation.  I need help, as hinted in https://github.com/openssl/openssl/pull/13436#issuecomment-729591924

levitte commented:
PING

mattcaswell commented on ssl/t1_lib.c:
```diff
@@ -3452,3 +3450,27 @@ size_t ssl_hmac_size(const SSL_HMAC *ctx)
+
+    if (EVP_PKEY_id(pkey) != EVP_PKEY_KEYMGMT) {
+        if (EVP_PKEY_base_id(pkey) == EVP_PKEY_EC) {
+            EC_KEY *ec = EVP_PKEY_get0_EC_KEY(pkey);
```
Noooooo!!!!!!!!!! All EC_KEY functions are about to be deprecated (#13139) we must not introduce new usage!

levitte replied:
I've updated this function to use existing `EVP_PKEY_CTX_get_group_name()`...  Muuuuch neater

mattcaswell commented on ssl/t1_lib.c:
```diff
@@ -3452,3 +3450,27 @@ size_t ssl_hmac_size(const SSL_HMAC *ctx)
+        char name[80] = { 0 };
+        size_t name_len;
+
+        if (EVP_PKEY_get_utf8_string_param((EVP_PKEY *)pkey,
```
We need to get this to work with legacy keys so that the first "if" branch is not necessary.

mattcaswell replied:
i.e. by exporting it to a provider if necessary

levitte replied:
> i.e. by exporting it to a provider if necessary

That's a *lot* of churn just to get the curve name.  While I sympathise with the purity standpoint, there's also something to be said about pragmatism.

levitte replied:
Something we *could* do is a legacy section in `EVP_PKEY_get_utf8_string_param()` that handles legacy keys...  The inverse of functions like `legacy_ctrl_to_param()` if you will...  *Much* less churn.

mattcaswell replied:
> Something we could do is a legacy section in EVP_PKEY_get_utf8_string_param() that handles legacy keys... The inverse of functions like legacy_ctrl_to_param() if you will... Much less churn.

That would be ok with me.

levitte replied:
For another time...

mattcaswell commented on ssl/t1_lib.c:
```diff
@@ -3452,3 +3450,27 @@ size_t ssl_hmac_size(const SSL_HMAC *ctx)
+        if (EVP_PKEY_get_utf8_string_param((EVP_PKEY *)pkey,
+                                           OSSL_PKEY_PARAM_GROUP_NAME,
+                                           name, sizeof(name), &name_len))
+            ret = EC_curve_name2nid(name);
```
I think this is the wrong approach. Rather than indexing on nid we could index on the TLS_GROUP_INFO entry in the ctx->group_list array. We should scan that list looking for a "realname" that matches the curve name we've just got from the provider.

levitte replied:
> I think this is the wrong approach. Rather than indexing on nid we could index on the TLS_GROUP_INFO entry in the ctx->group_list array. We should scan that list looking for a "realname" that matches the curve name we've just got from the provider.

Do we have all those details in place?  If not, that must happen first.  All in all, though, I completely agree with you.

mattcaswell replied:
> Do we have all those details in place? If not, that must happen first. All in all, though, I completely agree with you.

Yes. This is the basis of the capabilities work I did. We ask the providers what groups they support using the "TLS-GROUP" capability and populate `ctx->group_list` with everything we found.

levitte replied:
I've kept this to be for NIDs for now...  The rest is a much bigger refactor that I think needs to be done separately.

mattcaswell reviewed (changes requested):
I've not reviewed all of this PR...but I think we're heading down the wrong road with it

levitte commented:
Oh, and I rebased too

mattcaswell commented:
CIs don't like it.

mattcaswell commented on ssl/t1_lib.c:
```diff
@@ -3456,3 +3455,17 @@ size_t ssl_hmac_size(const SSL_HMAC *ctx)
+
+    if ((EVP_PKEY_is_a(pkey, "EC") || EVP_PKEY_is_a(pkey, "SM2"))
+        && ((ctx = EVP_PKEY_CTX_new((EVP_PKEY *)pkey, NULL)) != NULL
+            || EVP_PKEY_CTX_get_group_name(ctx, gname, sizeof(gname)) > 0))
```
Don't we need to modify `EVP_PKEY_CTX_get_group_name` to work with legacy keys? Because it doesn't right now.

levitte replied:
Oops, forgot about that. Will fix tomorrow

levitte commented:
After some longer discussion, primarily with @mattcaswell, I reworked this whole PR.  `evp_pkey_get_EC_KEY_curve_nid()` is still gone, that hasn't changed.  I've changed the PR description accordingly.

t8m commented on ssl/t1_lib.c:
```diff
@@ -3155,7 +3155,7 @@ static const SIGALG_LOOKUP *find_sig_alg(SSL *s, X509 *x, EVP_PKEY *pkey)
         if (lu->sig == EVP_PKEY_EC) {
 #ifndef OPENSSL_NO_EC
```
Should not it be possible to remove this ifndef now?

levitte replied:
Done

t8m commented on ssl/t1_lib.c:
```diff
@@ -3218,9 +3218,8 @@ int tls_choose_sigalg(SSL *s, int fatalerrs)
-                    curve =
-                        evp_pkey_get_EC_KEY_curve_nid(s->cert->pkeys[SSL_PKEY_ECC]
-                                                      .privatekey);
+                    curve = ssl_get_EC_curve_nid(s->cert->pkeys[SSL_PKEY_ECC]
```
Here it is in OPENSSL_NO_EC as well but probably does not need to anymore.

levitte replied:
Done, although I'm a little unsure about the consequences...

levitte commented on ssl/t1_lib.c:
```diff
@@ -3250,9 +3243,7 @@ int tls_choose_sigalg(SSL *s, int fatalerrs)
                     }
-#ifndef OPENSSL_NO_EC
                     if (curve == -1 || lu->curve == curve)
-#endif
```
@mattcaswell, could you double check that this makes sense?  I'm a bit apprehensive about that `curve == -1` check...

mattcaswell replied:
This all looks correct to me.

levitte replied:
Okie.  Thanks.

t8m commented on ssl/t1_lib.c:
```diff
@@ -3153,14 +3153,10 @@ static const SIGALG_LOOKUP *find_sig_alg(SSL *s, X509 *x, EVP_PKEY *pkey)
                                  : s->cert->pkeys[lu->sig_idx].privatekey;
 
         if (lu->sig == EVP_PKEY_EC) {
-#ifndef OPENSSL_NO_EC
```
There is one more ifndef OPENSSL_NO_EC above that needs to be dropped.

t8m commented on ssl/t1_lib.c:
```diff
@@ -1498,7 +1498,7 @@ int tls12_check_peer_sigalg(SSL *s, uint16_t sig, EVP_PKEY *pkey)
 
         /* For TLS 1.3 or Suite B check curve matches signature algorithm */
         if (SSL_IS_TLS13(s) || tls1_suiteb(s)) {
-            int curve = evp_pkey_get_EC_KEY_curve_nid(pkey);
```
And here it is enclosed in ifndef OPENSSL_NO_EC as well. I am sorry for taking you to this path. Maybe this should have been deferred to a separate PR.

levitte replied:
Yes, I know...  there's a call in `ssl/statem/statem_lib.c` as well, and quite frankly, I can't see the consequences clearly enough to dare fix them all unless someone can tell me for sure that I have nothing to worry about (@mattcaswell is my guru in this, but if you are confident that you see clearly enough, I'll trust you).

t8m replied:
I am afraid there might be some cascading effect. So probably better to defer this. I've opened #13616 to cover this in general.

levitte replied:
Cool.  Then I'll resolves these last comments unchanged.

levitte replied:
> I am afraid there might be some cascading effect.

I share the feeling.

Thank you, I'll resolve these remaining concerns unchanged, then.

levitte commented:
Rebased to fix merge conflicts

levitte commented on doc/man3/EVP_PKEY_get_group_name.pod:
```diff
@@ -0,0 +1,46 @@
+=head1 DESCRIPTION
+
+EVP_PKEY_get_group_name() fills in the group name of the I<pkey> into
+I<mdname>, up to at most I<mdname_sz> bytes including the ending NUL byte
```
Fixed

levitte commented on doc/man3/EVP_PKEY_get_group_name.pod:
```diff
@@ -0,0 +1,46 @@
+
+=head1 NOTES
+
+Among the standard OpenSSL key types, this is only support for DH, EC and
```
Fixed

mattcaswell commented:
CI failures look relevant

levitte commented:
> CI failures look relevant

Hopefully fixed

mattcaswell reviewed (approved):
LGTM

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but this PR has failing CI tests.  Once the tests pass it will get moved to 'approval: ready to merge' automatically, alternatively please review and set the label manually.

levitte commented:
Merged

a73a1892221e04ddb8ff9ec85ebaa48b5a853de6 EVP: constify the EVP_PKEY_get_*_param() argument |pkey|
88bddad42ee27483d153a0b0c0edd13b2b5fdbc0 EVP: Add EVP_PKEY_get_group_name() to extract the group name of a pkey
d8975dec0c3f41a491345f8a3c02612eaf8b30f7 TLS: Use EVP_PKEY_get_group_name() to get the group name
